<?xml version="1.0"?>

<!-- ##################################################### -->
<!--                   URDF DESCRIPTION                    -->
<!-- ##################################################### -->

<robot name="mapping_robot" xmlns:xacro="http://www.ros.org/wiki/xacro">


<!-- ##################################################### -->
<!--                 Geometry && Constants                 -->
<!-- ##################################################### -->

<!-- PI Constant -->
<xacro:property name="PI" value="3.14159265359"/>

<!-- Body Dimensions -->
<!-- R = side of robot body -->
<!-- H = height of robot body -->
<xacro:property name="R" value="0.3"/>
<xacro:property name="H" value="0.25"/>

<!-- Wheels Dimensions -->
<!-- r = wheels && caster radius -->
<!-- w = wheels width -->
<xacro:property name="r" value="0.034"/>
<xacro:property name="w" value="0.028"/>

<!-- Lengths to centers -->
<!-- s1 = length from c1 to c2 -->
<!-- s2 = length from c2 to c3 && c2 to c4 -->
<!-- s3 = length from c2 to c5 -->
<xacro:property name="s1" value="${H/2+r}"/>
<xacro:property name="s2" value="${R/2.0}"/>
<xacro:property name="s3" value="${R/2.0-0.03}"/>


<!-- ##################################################### -->
<!--                    Mass && Inertia                    -->
<!-- ##################################################### -->

<!-- Material density -->
<!-- Body material density -->
<xacro:property name="d1" value="2710.0"/>
<!-- Wheel material density -->
<xacro:property name="d2" value="2710.0"/>
<!-- Caster material density -->
<xacro:property name="d3" value="2710.0"/>

<!-- ################### MASS, VOLUME, CENTER OF MASS EQU ################### -->
<!-- Volumes -->
<!-- Volumes cylinders: V = PI*w*r^2 -->
<!-- Volumes sphere: V = (4/3)*PI*r^3 -->

<!-- Masses -->
<!-- Masses of cylinders: M = DENSITY*V -->
<!-- Masses of sphere: M = DENSITY*V -->

<!-- Momments of inertia -->
<!-- Momments of inertia of cylinders:
(
Ix = (1/12)*M*((3*r^2)+w^2)
Iy = (1/12)*M*((3*r^2)+w^2)
Iz = (1/2)*M*r^2
) -->
<!-- Momments of inertia of sphere:
(
Ix = (2/5)*M*r^2
Iy = (2/5)*M*r^2
Iz = (2/5)*M*r^2
) -->

<!-- MASSES -->
<!-- Body -->
<xacro:property name="m1" value="${d1*H*R*R}"/>
<!-- Wheel -->
<xacro:property name="m2" value="${d2*PI*w*r*r}"/>
<!-- Caster -->
<xacro:property name="m3" value="${d3*(4.0/3.0)*PI*r*r*r}"/>

<!-- Momments of inertia Body -->
<xacro:property name="Ix_body" value="${(1.0/12.0)*m1*(R*R+H*H)}"/>
<xacro:property name="Iy_body" value="${(1.0/12.0)*m1*(R*R+H*H)}"/>
<xacro:property name="Iz_body" value="${(1.0/12.0)*m1*(R*R+R*R)}"/>

<!-- Momments of inertia Wheel -->
<xacro:property name="I_wheel" value="${(1.0/12.0)*m2*(3.0*r*r+w*w)}"/>
<xacro:property name="Iz_wheel" value="${0.5*m2*r*r}"/>

<!-- Momments of inertia Caster -->
<xacro:property name="I_caster" value="${0.4*m3*r*r}"/>


<!-- Center of mass body (c2) -->
<xacro:macro name="inertia_body">
    <inertial>
    <origin rpy="0 0 0" xyz="0 0 ${s1}"/>
    <mass value="${m1}"/>
    <inertia ixx="${Ix_body}" ixy="0.0" ixz="0.0" iyy="${Iy_body}" iyz="0.0" izz="${Iz_body}"/>
    </inertial>
</xacro:macro>

<!-- Center of mass wheel (c3-c4) -->
<xacro:macro name="inertia_wheel">
    <inertial>
    <origin rpy="1.5707963 0 0" xyz="0 0 0"/>
    <mass value="${m2}"/>
    <inertia ixx="${I_wheel}" ixy="0.0" ixz="0.0" iyy="${I_wheel}" iyz="0.0" izz="${Iz_wheel}"/>
    </inertial>
</xacro:macro>

<!-- Center of mass caster (c5) -->
<xacro:macro name="inertia_caster">
    <inertial>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <mass value="${m3}"/>
    <inertia ixx="${I_caster}" ixy="0.0" ixz="0.0" iyy="${I_caster}" iyz="0.0" izz="${I_caster}"/>
    </inertial>
</xacro:macro>


<!-- ##################################################### -->
<!--               Include additional files                -->
<!-- ##################################################### -->

<xacro:include filename="$(find RL_robot)/model/robot.gazebo"/>


<!-- ##################################################### -->
<!--                    Links && Joints                    -->
<!-- ##################################################### -->

<!-- a neccessery link for gazebo (base footprint = c1, body link = c2) -->
<link name="base_footprint"> </link>

<joint name="body_link_joint" type="fixed">
    <parent link="base_footprint"/>
    <child link="body_link"/>
</joint>

<!-- Body link -->
<link name="body_link">
    <visual>
        <geometry>
            <box size="${R} ${R} ${H}"/>
        </geometry>
        <origin rpy="0 0 0" xyz="0 0 ${s1}"/>
    </visual>
    <collision>
        <geometry>
            <box size="${R} ${R} ${H}"/>
        </geometry>
        <origin rpy="0 0 0" xyz="0 0 ${s1}"/>
    </collision>
    <xacro:inertia_body />
</link>

<!-- ####### Wheels ####### -->
<!-- wheel1 -->
<joint name="wheel1_joint" type="continuous">
    <parent link="body_link"/>
    <child link="wheel1_link"/>
    <origin rpy="0 0 0" xyz="${-s3} ${-s2} ${r}"/>
    <axis xyz="0 1 0"/>
    <limit effort="50000" velocity="10"/>
    <dynamics damping="1.0" friction="1.0"/>
</joint>

<!-- wheel1 link -->
<link name="wheel1_link">
    <visual>
        <geometry>
            <cylinder length="${w}" radius="${r}"/>
        </geometry>
        <origin rpy="1.5707963 0 0" xyz="0 0 0"/>
    </visual>
    <collision>
        <geometry>
            <cylinder length="${w}" radius="${r}"/>
        </geometry>
        <origin rpy="1.5707963 0 0" xyz="0 0 0"/>
    </collision>
    <xacro:inertia_wheel />
</link>

<!-- wheel2 -->
<joint name="wheel2_joint" type="continuous">
    <parent link="body_link"/>
    <child link="wheel2_link"/>
    <origin rpy="0 0 0" xyz="${-s3} ${s2} ${r}"/>
    <axis xyz="0 1 0"/>
    <limit effort="50000" velocity="10"/>
    <dynamics damping="1.0" friction="1.0"/>
</joint>

<!-- wheel2 link -->
<link name="wheel2_link">
    <visual>
        <geometry>
            <cylinder length="${w}" radius="${r}"/>
        </geometry>
        <origin rpy="1.5707963 0 0" xyz="0 0 0"/>
    </visual>
    <collision>
        <geometry>
            <cylinder length="${w}" radius="${r}"/>
        </geometry>
        <origin rpy="1.5707963 0 0" xyz="0 0 0"/>
    </collision>
    <xacro:inertia_wheel />
</link>


<!-- ####### Casters ####### -->
<!-- caster1 -->
<joint name="caster1_joint" type="fixed">
    <parent link="body_link"/>
    <child link="caster1_link"/>
</joint>

<!-- caster1 link -->
<link name="caster1_link">
    <visual>
        <geometry>
            <sphere radius="${r}"/>
        </geometry>
        <origin rpy="0 0 0" xyz="${s3} ${s3} ${r}"/>
    </visual>
    <collision>
        <geometry>
            <sphere radius="${r}"/>
        </geometry>
        <origin rpy="0 0 0" xyz="${s3} ${s3} ${r}"/>
    </collision>
    <xacro:inertia_caster />
</link>

<!-- caster2 -->
<joint name="caster2_joint" type="fixed">
    <parent link="body_link"/>
    <child link="caster2_link"/>
</joint>

<!-- caster2 link -->
<link name="caster2_link">
    <visual>
        <geometry>
            <sphere radius="${r}"/>
        </geometry>
        <origin rpy="0 0 0" xyz="${s3} ${-s3} ${r}"/>
    </visual>
    <collision>
        <geometry>
            <sphere radius="${r}"/>
        </geometry>
        <origin rpy="0 0 0" xyz="${s3} ${-s3} ${r}"/>
    </collision>
    <xacro:inertia_caster />
</link>

<!-- RPLIDAR sensor -->
<joint name="lidar_joint" type="fixed">
    <parent link="body_link"/>
    <child link="lidar_link"/>
</joint>

<!-- rplidar link -->
<link name="lidar_link">
    <visual>
        <geometry>
            <cylinder length="${2*w}" radius="${r}"/>
        </geometry>
        <origin rpy="0 0 0" xyz="${-s3} 0 ${(H/2)+w+s1}"/>
    </visual>
    <collision>
        <geometry>
            <cylinder length="${2*w}" radius="${r}"/>
        </geometry>
        <origin rpy="0 0 0" xyz="${-s3} 0 ${(H/2)+w+s1}"/>
    </collision>
    <xacro:inertia_wheel />
</link>

<!-- ##################################################### -->
<!-- END                                                   -->
<!-- ##################################################### -->

</robot>